### PR TITLE
[IO] Pass client builder to debezium database history

### PR DESCRIPTION
**Pull request by sijie**
an alternative approach for https://github.com/apache/pulsar/pull/11251

### CODE REVIEW

File: pulsar-io/debezium/core/src/main/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistory.java [hunk 1]
```diff
@@ -61,26 +66,34 @@
         .withValidation(Field::isRequired);
 
     public static final Field SERVICE_URL = Field.create(CONFIGURATION_FIELD_PREFIX_STRING + "pulsar.service.url")
-        .withDisplayName("Pulsar broker addresses")
+        .withDisplayName("Pulsar service url")
         .withType(Type.STRING)
         .withWidth(Width.LONG)
         .withImportance(Importance.HIGH)
         .withDescription("Pulsar service url")
         .withValidation(Field::isRequired);
```
Comment (nlu90) [inline pulsar-io/debezium/core/src/main/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistory.java:74]: we can set either this or the following field. so they might not be `required` now.

**Review by nlu90 — COMMENTED** [review]
LGTM for functionality purpose.

I'm just a little concerned about:
1. the new exposed API.
2. the security of passing this object string around.

If you can help explain it's secure to do so, that will be great.

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java [hunk 2]
```diff
@@ -201,4 +202,14 @@
     default PulsarClient getPulsarClient() {
         throw new UnsupportedOperationException("not implemented");
     }
+
+    /**
+     * Get the pulsar client builder.
```
Comment (eolivelli) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java:207]: can we add something like "This Builder is pre-configured with Connection parameters, Authentication credentials and other context values.
You can use this Builder in order to connect to the local Pulsar cluster.
The Function will be responsible for disposing any PulsarClient created by this builder."

File: pulsar-io/debezium/core/src/main/java/org/apache/pulsar/io/debezium/DebeziumSource.java [hunk 3]
```diff
@@ -90,6 +95,16 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
         setConfigIfNull(config, PulsarKafkaWorkerConfig.OFFSET_STORAGE_TOPIC_CONFIG,
             topicNamespace + "/" + sourceName + "-" + DEFAULT_OFFSET_TOPIC);
 
+        ClientBuilder clientBuilder = sourceContext.getPulsarClientBuilder();
+        ByteArrayOutputStream bao = new ByteArrayOutputStream();
+        try (ObjectOutputStream oos = new ObjectOutputStream(bao)) {
```
Comment (eolivelli) [inline pulsar-io/debezium/core/src/main/java/org/apache/pulsar/io/debezium/DebeziumSource.java:100]: what about moving this code to some utility method?

**Review by eolivelli — COMMENTED** [review]
Overall the approach looks good to me, it is a great idea indeed.


One question related to security:
with this change the Function code is able to easily access the Authentication information (because it can Serialize the Builder and then read  the credentials).
Before this change (with the previous approach) the Function was not able to access the credentials but only to use the PulsarClient, for instance a JWT token could not be stolen. Now you can deploy the Function (the auth information is set by the Admin who deploys the function, not by the author of the function) and the Function can get the credentials and send them outside the cluster.

is it something we should care about ?

**Review by eolivelli — APPROVED** [review]

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java [hunk 4]
```diff
@@ -194,11 +195,27 @@
...
+
+    /**
+     * Get the pre-configured pulsar client builder.
+     *
+     * You can use this Builder to setup client to connect to the Pulsar cluster.
+     * The Function will be responsible for disposing any PulsarClient created by this builder.
+     *
+     * @return the instance of pulsar client builder.
+     */
+    default ClientBuilder getPulsarClientBuilder() {
```
Comment (sijie) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java:217]: @nlu90 You need to implement this method in the sub-classes.

Comment (nlu90) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java:217]: working on it

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java (hunk 2 above)
Comment (nlu90) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/BaseContext.java]: After a second thought, I think User should be responsible for cleaning any client created by using the ClientBuilder.

The Function Runtime is only responsible for closing the client get via `getPulsarClient()`.

File: pulsar-io/debezium/core/src/test/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistoryTest.java [hunk 5]
```diff
@@ -22,6 +22,7 @@
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertTrue;
 
+import com.google.api.client.util.Base64;
```
Comment (eolivelli) [inline pulsar-io/debezium/core/src/test/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistoryTest.java:25]: please use java.util.Base64

**Review by eolivelli — CHANGES_REQUESTED** [review]
I have reviewed the patch again.
I left two comments, PTAL

File: pulsar-io/debezium/core/src/test/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistoryTest.java (hunk 5 above)
Comment (nlu90) [reply pulsar-io/debezium/core/src/test/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistoryTest.java:25]: 👍🏼

**Comment by hangc0276** [comment]
@eolivelli Cloud you please help review this PR again, thanks.

**Comment by eolivelli** [comment]
There are still some unaddressed comments of mine:
- do not use google base64
- create a utility method to serialize/deserialize the ClientBuilder

File: pulsar-io/debezium/core/src/main/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistory.java [hunk 6]
```diff
@@ -29,13 +29,19 @@
...
 import io.debezium.relational.history.HistoryRecord;
 import io.debezium.relational.history.HistoryRecordComparator;
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
+import java.io.InputStream;
+import java.io.ObjectInputStream;
 import java.util.UUID;
 import java.util.function.Consumer;
 import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.codec.binary.Base64;
```
Comment (eolivelli) [inline pulsar-io/debezium/core/src/main/java/org/apache/pulsar/io/debezium/PulsarDatabaseHistory.java:39]: Please use Java util

**Review by eolivelli — APPROVED** [review]
looks great to me

**Comment by eolivelli** [comment]
integration tests failed, regarding Debezium, please check

```
testDebeziumMySqlSourceAvro(org.apache.pulsar.tests.integration.io.sources.debezium.PulsarDebeziumSourcesTest)  Time elapsed: 56.866 s  <<< FAILURE!
java.lang.AssertionError: expected [1] but found [0]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.tests.integration.io.sources.PulsarIOSourceRunner.getSourceStatus(PulsarIOSourceRunner.java:210)
	at org.apache.pulsar.tests.integration.io.sources.debezium.PulsarIODebeziumSourceRunner.lambda$testSource$0(PulsarIODebeziumSourceRunner.java:76)
	at net.jodah.failsafe.Functions.lambda$toSupplier$8(Functions.java:236)
	at net.jodah.failsafe.Functions.lambda$get$0(Functions.java:47)
	at net.jodah.failsafe.RetryPolicyExecutor.lambda$supply$0(RetryPolicyExecutor.jav
```

**Comment by hangc0276** [comment]
move to 2.8.2.

**Comment by eolivelli** [comment]
@hangc0276 this is a new API, probably it won't be delivered in 2.8.x. I suggest to remove the 2.8.2 label.

in point releases we should not add new APIs

**Review by freeznet — APPROVED** [review]

**Comment by sijie** [comment]
@nlu90 Can you rebase the PR to latest master? I merged the upgrade debezium version change.

File: pulsar-io/debezium/core/pom.xml [hunk 7]
```diff
@@ -62,12 +62,6 @@
       <version>${kafka-client.version}</version>
     </dependency>
 
-    <dependency>
-      <groupId>${project.groupId}</groupId>
-      <artifactId>pulsar-client-original</artifactId>
```
Comment (eolivelli) [inline pulsar-io/debezium/core/pom.xml:67]: This is very interesting.
it looks like that we do not need to import the pulsar-client-original anymore

**Review by eolivelli — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**